src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1542,9 +1542,31 @@ private BoundExpression InferTypeForDiscardAssignment(BoundDiscardExpression op1
                 {
                     if (isRef)
                     {
+                        // https://github.com/dotnet/csharplang/blob/main/proposals/low-level-struct-improvements.md#rules-ref-reassignment
+                        // For a ref reassignment in the form ...
+                        // 1. x.e1 = ref e2: where x is safe-to-escape at least return only then e2 must have ref-safe-to-escape at least as large as x
+                        // 2. e1 = ref e2: where e1 is a ref local or ref parameter then e2 must have a safe-to-escape equal to safe-to-escape for e1 and
+                        //    e2 must have ref-safe-to-escape at least as large as ref-safe-to-escape of the ref-safe-to-escape of e1
+
+                        bool reportRefEscape;
                         var leftEscape = GetRefEscape(op1, LocalScopeDepth);
                         var rightEscape = GetRefEscape(op2, LocalScopeDepth);
                         if (leftEscape < rightEscape)
+                        {
+                            reportRefEscape = true;
+                        }
+                        else if (op1.Kind is BoundKind.Local or BoundKind.Parameter &&
+                            (leftEscape = GetValEscape(op1, LocalScopeDepth)) != (rightEscape = GetValEscape(op2, LocalScopeDepth)))
```

Comment:
Here, we're checking for equality, but the error message complains about one being "narrower". We probably need distinct error messages.